### DISCUSSION TITLE
Media upload with pre-signed URL

### DISCUSSION THREAD

**Discussion by janbuecker** [shopware/platform#3159] [RFC / Proposal]
**Effort**: low

### Description: 

Media files that are being uploaded, always pass the controller and the backend copies the stream to the desired path in the file system. That means, that every file crosses the application server in a whole. In scenarios, where you have an object storage like AWS S3, you double the traffic, because the user uploads the file to the application server, which then uploads the file to the remote file system. 

Many object storage services offer "pre-signed URLs". You call the service to generate a one-time token based on parameters like the path and get a URL including that token in return. You can then use that URL to let the user upload the file directly to the remote file system and skip the traffic to the application server.

### Benefits

- Reduced costs with remote file-system implementations
- Less load on the application server

### Break strategy

For file systems with no support for pre-signed URLs (e.g. local), you could provide a controller endpoint, which simulates this behavior and another one, which generates the pre-signed URL. So in turn, deprecate the old endpoint and remove it with Version 6.x.

**Comment by OliverSkroblin** [comment]
Hey @janbuecker,
your Idea sounds great and i think many other cloud services would be happy with the solution. Could you perhaps go into a bit more detail about the affected code lines? 
Of course, it doesn't have to be a complete concept, but it seems that you have already dealt with the topic a bit more and have certainly already identified the affected code passages. Therefore, it would be cool if you could include a few of these places in the proposal to clarify which areas are involved in the platform.

**Comment by shyim** [comment]
I am not sure about the traffic improvements. We have to download the files afterwards to generate thumbnails or analyst the uploaded file. I don't really see here any benefit 😅

**Comment by janbuecker** [comment]
@OliverSkroblin ok

@shyim Yes, but this is true for images only. You could even delegate the generation of thumbnails to another service and work with events. In case of s3, you can automatically trigger a lambda on upload and handle thumbnails there. Inter-VPC traffic is much cheaper than through the internet. It’s totally fine to have it all on the app server but with this change you enable such possibilities described above.

**Comment by OliverSkroblin** [comment]
Should be realized with [this commit](https://github.com/shopware/platform/commit/0fa3f83e2a97d9c1a4c4c3b63b34bb2331577177)
